

<!--
 * @version:
 * @Author: steven
 * @Date: 2020-06-11 21:19:10
 * @LastEditors: steven
 * @LastEditTime: 2020-06-11 22:44:23
 * @Description:
-->
1. I'm not very clear. Maybe second derivative is more easy to overflow the scope of storages.
2.
```python
y.backward()
```
RuntimeError: Trying to backward through the graph a second time, but the buffers have already been freed. Specify retain_graph=True when calling backward the first time.
---
```python
y = 2 * torch.dot(x, x)
y
y.backward(retain_graph=True)
y.backward(retain_graph=True)
```
no Error!
3. Error happened!
```python
# matrix
a = torch.randn(20, requires_grad=True).reshape(5, 4)
d = f(a)
d.backward()
```

RuntimeError: grad can be implicitly created only for scalar outputs

---

analyze:

source code:~\.conda\envs\pytorch\lib\site-packages\torch\autograd\__init__.py in _make_grads(outputs, grads)

rows:from 32 to 35 :
```python
        elif grad is None: # why grad is None? Does f(a) not support matrix?
            if out.requires_grad:
                if out.numel() != 1:
                    raise RuntimeError("grad can be implicitly created only for scalar outputs")
```
4. TODO:
5. I'm trying to import plot in 2.4, so I convert the "2-4.ipynb" to "two-four.py".

But when I import it, it just happens as follow. How to fix it?
```python
from ..d2l import torch as d2l
from IPython import display
import numpy as np
from .two_four import plot
```

ImportError                               Traceback (most recent call last)
 in
----> 1 from ..d2l import torch as d2l
      2 from IPython import display
      3 import numpy as np
      4 from .two_four import plot

ImportError: attempted relative import with no known parent package

TODO:
